tests/MongoDB.Driver.Core.Tests/Core/Authentication/ScramSha1AuthenticatorTests.cs

```diff
@@ -189,27 +205,47 @@ public void Constructor_should_throw_an_ArgumentNullException_when_credential_is
             }
 
-            sentMessages[0].Should().Be(
-                $"{{opcode: \"query\", " +
+            var saslStartMessage = BsonDocument.Parse(
+                @"{ opcode: ""query""," +
                 $"  requestId: {actualRequestIds[0]}, " +
-                $"  database: \"source\", collection: \"$cmd\", batchSize: -1, slaveOk: true, " +
-                $"  query: {{saslStart: 1, mechanism: \"SCRAM-SHA-1\", " +
+                @"  database: ""source"",
+                    collection: ""$cmd"",
+                    batchSize: -1,
+                    slaveOk: true,
+                    query: { saslStart: 1,
+                             mechanism: ""SCRAM-SHA-1""," +
                 $"           payload: new BinData(0, \"biwsbj11c2VyLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdM\")" +
                 @"           options: { skipEmptyExchange: true }}}");
-            sentMessages[1].Should().Be("{opcode: \"query\", requestId: " + actualRequestIds[1] + ", database: \"source\", collection: \"$cmd\", batchSize: -1, slaveOk: true, query: {saslContinue: 1, conversationId: 1, payload: new BinData(0, \"Yz1iaXdzLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdMSG8rVmdrN3F2VU9LVXd1V0xJV2c0bC85U3JhR01IRUUscD1NQzJUOEJ2Ym1XUmNrRHc4b1dsNUlWZ2h3Q1k9\")}}");
-            if (useLongAuthentication)
-            {
-                sentMessages[2].Should().Be(
+
+            var saslContinueMessage = BsonDocument.Parse(
+                @"{ opcode: ""query""," +
+                $"  requestId: {(useSpeculativeAuthenticate ? actualRequestIds[0] : actualRequestIds[1])}," +
+                @"  database: ""source"",
+                    collection: ""$cmd"",
+                    batchSize: -1,
+                    slaveOk: true,
+                    query: { saslContinue: 1,
+                             conversationId: 1, " +
+                $"           payload: new BinData(0, \"Yz1iaXdzLHI9ZnlrbytkMmxiYkZnT05Sdjlxa3hkYXdMSG8rVmdrN3F2VU9LVXd1V0xJV2c0bC85U3JhR01IRUUscD1NQzJUOEJ2Ym1XUmNrRHc4b1dsNUlWZ2h3Q1k9\")}}}}");
+             var saslOptionalFinalMessage = new Lazy<BsonDocument>(() => BsonDocument.Parse(
                     @"{opcode: ""query""," +
-                    $" requestId: {actualRequestIds[2]}," +
+                    $" requestId: {(useSpeculativeAuthenticate ? actualRequestIds[1] : actualRequestIds[2])}," +
                     @" database: ""source"",
                        collection: ""$cmd"",
                        batchSize: -1,
                        slaveOk: true,
-                       query: {saslContinue: 1,
-                               conversationId: 1, " +
-                    $"         payload: new BinData(0, \"\")}}}}");
-            }
+                       query: { saslContinue: 1,
+                                conversationId: 1, " +
+                    $"          payload: new BinData(0, \"\")}}}}"));
+
+            var expectedMessages = new[]
```

Comment:
I think Dmitry used mixed tabs and spaces in his example but in general I would agree. The continuation lines need to be indented.

I also find it kind of weird that you include nulls in the array that then have to be removed with a Where clause. You are obfuscating through functiofication.

I would use:

    var expectedMessage = new List<BsonDocument>();
    if (!useSpeculativeAuthenticate)
    {
        expectedMessages.Add(saslStartMessage);
    }
    expectedMessages.Add(saslContinueMessage);
    if (useLongAuthentication )
    {
        var saslFinalMessage = BsonDocument.Parse(...); // no need for Lazy now
        expectedMessages.Add(saslFinalMessage);
    }

Reply:
I prefer the `Where` version, since it does result in a mildly smaller diff, but admit that price of my code golfing is a slightly weird Lazy. To move things along, I'm fine with taking your version :)